### ISSUE TITLE
Faraday can't make requests to Akamai URLs

### ISSUE THREAD

**Issue by tijmenb** [lostisland/faraday#1598]
## Basic Info

* Faraday Version: 1.10.4
* Ruby Version: `ruby 3.3.4 (2024-07-09 revision be1089c8ec) [arm64-darwin24]`

## Issue description

We have a system that fetches content from a third-party system. This works for most URLs, but Faraday doesn't seem to be able to make correct the GET requests to Akamai URLs that look like this:

https://bcbolt446c5271-a.akamaihd.net/media/v1/text/vtt/clear/5784545058001/35a80311-505b-45c4-a498-d6e7ad005d21/9284a643-1722-4441-b04e-270caf143754/text.vtt?akamai_token=exp=1730832553~acl=/media/v1/text/vtt/clear/5784545058001/35a80311-505b-45c4-a498-d6e7ad005d21/9284a643-1722-4441-b04e-270caf143754/text.vtt*~hmac=40272eaf84d5f966fdcbe18f68eff327192c6785b332b974983bea916dea520b  

## Steps to reproduce

First, confirm that the above URL works in a browser and that it works using cURL:

```sh
curl -i "https://bcbolt446c5271-a.akamaihd.net/media/v1/text/vtt/clear/5784545058001/35a80311-505b-45c4-a498-d6e7ad005d21/9284a643-1722-4441-b04e-270caf143754/text.vtt?akamai_token=exp=1730832553~acl=/media/v1/text/vtt/clear/5784545058001/35a80311-505b-45c4-a498-d6e7ad005d21/9284a643-1722-4441-b04e-270caf143754/text.vtt*~hmac=40272eaf84d5f966fdcbe18f68eff327192c6785b332b974983bea916dea520b"
```

Returns:

```
HTTP/1.1 200 OK
...
```

Now make a GET request using Faraday, and observe that Akamai returns a 403:

```rb
irb(main):001> Faraday.new.get("https://bcbolt446c5271-a.akamaihd.net/media/v1/text/vtt/clear/5784545058001/35a80311-505b-45c4-a498-d6e7ad005d21/9284a643-1722-4441-b04e-270caf143754/text.vtt?akamai_token=exp=1730832553~acl=/media/v1/text/vtt/clear/5784545058001/35a80311-505b-45c4-a498-d6e7ad005d21/9284a643-1722-4441-b04e-270caf143754/text.vtt*~hmac=40272eaf84d5f966fdcbe18f68eff327192c6785b332b974983bea916dea520b")
=> 
#<Faraday::Response:0x000000013ec3e648
 @env=
  #<struct Faraday::Env
   method=:get,
   request_body=nil,
   url=
    #<URI::HTTPS https://bcbolt446c5271-a.akamaihd.net/media/v1/text/vtt/clear/5784545058001/35a80311-505b-45c4-a498-d6e7ad005d21/9284a643-1722-4441-b04e-270caf143754/text.vtt?akamai_token=exp%3D1730832553~acl%3D%2Fmedia%2Fv1%2Ftext%2Fvtt%2Fclear%2F5784545058001%2F35a80311-505b-45c4-a498-d6e7ad005d21%2F9284a643-1722-4441-b04e-270caf143754%2Ftext.vtt%2A~hmac%3D40272eaf84d5f966fdcbe18f68eff327192c6785b332b974983bea916dea520b>,
   request=#<struct Faraday::RequestOptions params_encoder=nil, proxy=nil, bind=nil, timeout=nil, open_timeout=nil, read_timeout=nil, write_timeout=nil, boundary=nil, oauth=nil, context=nil, on_data=nil>,
   request_headers={"User-Agent"=>"Faraday v1.10.4"},
   ssl=#<struct Faraday::SSLOptions verify=true, ca_file=nil, ca_path=nil, verify_mode=nil, cert_store=nil, client_cert=nil, client_key=nil, certificate=nil, private_key=nil, verify_depth=nil, version=nil, min_version=nil, max_version=nil>,
   parallel_manager=nil,
   params=nil,
   response=#<Faraday::Response:0x000000013ec3e648 ...>,
   response_headers=
    {"server"=>"AkamaiGHost",
     "mime-version"=>"1.0",
     "content-type"=>"text/html",
     "content-length"=>"280",
     "expires"=>"Tue, 05 Nov 2024 12:38:41 GMT",
     "date"=>"Tue, 05 Nov 2024 12:38:41 GMT",
     "connection"=>"keep-alive",
     "akamai-mon-iucid-del"=>"1315351",
     "alt-svc"=>"h3-Q050=\":443\"; ma=93600,quic=\":443\"; ma=93600; v=\"46,43\"",
     "access-control-allow-methods"=>"GET,HEAD,OPTIONS",
     "access-control-expose-headers"=>"Server,Range,Content-Length,Content-Range,Akamai-Mon-Iucid-Ing,Akamai-Mon-Iucid-Del,Akamai-Request-BC",
     "access-control-allow-headers"=>"X-Requested-With,Origin,Range,Accept-Encoding,Referer,CMCD-Request,CMCD-Object,CMCD-Status,CMCD-Session",
     "access-control-allow-origin"=>"*"},
   status=403,
   reason_phrase="Forbidden",
   response_body=
    "<HTML><HEAD><TITLE>Error</TITLE></HEAD><BODY>\nAn error occurred while processing your request.<p>\nReference&#32;&#35;199&#46;ec0e6168&#46;1730810321&#46;8375614\n<P>https&#58;&#47;&#47;errors&#46;edgesuite&#46;net&#47;199&#46;ec0e6168&#46;1730810321&#46;8375614</P>\n</BODY></HTML>\n">,
 @on_complete_callbacks=[]>
```

The 403 is caused by the URL having an invalid `akamai_token`, because Faraday escapes the values in the token:

```diff 
- https://bcbolt446c5271-a.akamaihd.net/media/v1/text/vtt/clear/5784545058001/35a80311-505b-45c4-a498-d6e7ad005d21/9284a643-1722-4441-b04e-270caf143754/text.vtt?akamai_token=exp=1730832553~acl=/media/v1/text/vtt/clear/5784545058001/35a80311-505b-45c4-a498-d6e7ad005d21/9284a643-1722-4441-b04e-270caf143754/text.vtt*~hmac=40272eaf84d5f966fdcbe18f68eff327192c6785b332b974983bea916dea520b
+ https://bcbolt446c5271-a.akamaihd.net/media/v1/text/vtt/clear/5784545058001/35a80311-505b-45c4-a498-d6e7ad005d21/9284a643-1722-4441-b04e-270caf143754/text.vtt?akamai_token=exp%3D1730832553~acl%3D%2Fmedia%2Fv1%2Ftext%2Fvtt%2Fclear%2F5784545058001%2F35a80311-505b-45c4-a498-d6e7ad005d21%2F9284a643-1722-4441-b04e-270caf143754%2Ftext.vtt%2A~hmac%3D40272eaf84d5f966fdcbe18f68eff327192c6785b332b974983bea916dea520b
```

Here's a bit of test code that disables escaping and makes it work for Akamai, for illustration:

```rb
module AkamaiURLEncoder
  def self.encode(params)
    params.map { |k, v| "#{k}=#{v}" }.join("&")
  end

  def self.decode(query)
    {
      akamai_token: query.split("akamai_token=").last,
    }
  end
end

client = Faraday.new do |f|
  f.options.params_encoder = AkamaiURLEncoder
end

client.get("https://bcbolt446c5271-a.akamaihd.net/media/v1/text/vtt/clear/5784545058001/35a80311-505b-45c4-a498-d6e7ad005d21/9284a643-1722-4441-b04e-270caf143754/text.vtt?akamai_token=exp=1730832553~acl=/media/v1/text/vtt/clear/5784545058001/35a80311-505b-45c4-a498-d6e7ad005d21/9284a643-1722-4441-b04e-270caf143754/text.vtt*~hmac=40272eaf84d5f966fdcbe18f68eff327192c6785b332b974983bea916dea520b")
```

Which happily returns a 200.

## Fix

I'm unsure if this is an issue in Faraday, or something that should be solved elsewhere. Akamai seems to reject encoded URL params, which seems odd (the whole `akamai_token` format seems weird). Should I be using a different method for requesting URLs that are hard to parse?

**Comment by iMacTia** [comment]
Hi @tijmenb and thank you for opening such a detailed issue.
Appreciate this may be surprising behaviour in your particular use-case, but Faraday is just doing what most people would expect in this case by escaping the characters in the value of the query-string param
> akamai_token=exp=1730832553~acl=/media/v1/text/vtt/clear/5784545058001/35a80311-505b-45c4-a498-d6e7ad005d21/9284a643-1722-4441-b04e-270caf143754/text.vtt*~hmac=40272eaf84d5f966fdcbe18f68eff327192c6785b332b974983bea916dea520b

Here, `akamai_token` is the key and *everything else* (i.e. between `=` and `&` or the end of the URL) is considered the value. Query-string values are escaped according to [RFC3986](https://datatracker.ietf.org/doc/html/rfc3986).

So in this case, I believe Faraday is doing the right thing, and instead the Akamai server is expecting a non standard-compliant URL.
The solution when dealing with this kind of servers, as you already pointed out, is to use a custom encoder. This type of scenario is also why we make this component configurable.

As much as I'd like for Faraday to be compatible with any server out-of-the-bow the web is just too big and varied to make this possible.

**Comment by tijmenb** [comment]
Thank you @iMacTia! I agree, Akamai is behaving weirdly here, and I understand the considerations.

I'm wondering about the expectations though, we did find it unexpected that Faraday escapes params, even when a full URL is provided. I've checked a few other Ruby HTTP clients/libraries, and I haven't actually been able to find a client that behaves like Faraday. The following requests all work and do not escape the params:

- `HTTParty.get(url)`
- `URI(url).read`
- `HTTP.get(url)`
- `Net::HTTP.get(URI(url))`
- `RestClient.get(url)`

Does that change the thinking at all?

To put it another way, I _would_ expect that params are escaped when doing `Faraday.new.get("http://example.com", foo: "bar=baz")`, but not `Faraday.new.get("http://example.com?foo=bar=baz")`.

I'll have a think on how we should write our custom encoder. We can't use a Akamai-specific one because the system also handles other user-provided URLs.

**Comment by iMacTia** [comment]
> The following requests all work and do not escape the params
> Does that change the thinking at all?

I'd say the difference in behaviour is understandable considering Faraday is not a pure HTTP client but a layer on top of them, so it makes sense for it to be "smarter" by default. I'm sorry the extra "smartness" didn't really help in your case 😓
And even if we did decide to align with HTTP clients, unfortunately changing the default behaviour is out of the question I'm afraid, as that would break the behaviour of all existing usages of Faraday.

The only remaining option would therefore be to make it easy for your to "opt-in" to a simplified behaviour that mimics the HTTP clients. To that extent, if all you want is to not have escaping, then this might work:

```ruby
CustomEncoder = Faraday::FlatParamsEncoder.dup
CustomEncoder.module_eval do
  def self.escape(str) = str
  def self.unescape(str) = str
end
```

I wish it was easier but the way `Faraday::FlatParamsEncoder` is defined doesn't make things easy 😄 
I quickly tried this on `irb` and it seems to work?
